server/src/main/java/org/elasticsearch/cluster/routing/allocation/NodeAllocationService.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.routing.allocation;
+
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.service.ClusterService;
+
+public class NodeAllocationService {
+
+    private final ClusterService clusterService;
+    private final WriteLoadForecaster writeLoadForecaster;
+
+    public NodeAllocationService(ClusterService clusterService, WriteLoadForecaster writeLoadForecaster) {
+        this.clusterService = clusterService;
+        this.writeLoadForecaster = writeLoadForecaster;
+    }
+
+    public NodeAllocationStats stats(String localNodeId) {
+        var state = clusterService.state();
+        var node = state.getRoutingNodes().node(localNodeId);
+
+        double forecastedWriteLoad = 0.0;
+        long forecastedDiskUsage = 0;
+        for (ShardRouting shardRouting : node) {
+            IndexMetadata indexMetadata = state.metadata().getIndexSafe(shardRouting.index());
+            forecastedWriteLoad += writeLoadForecaster.getForecastedWriteLoad(indexMetadata).orElse(0.0);
+            forecastedDiskUsage += indexMetadata.getForecastedShardSizeInBytes().orElse(0);// TODO fallback?
+        }
+
+        return new NodeAllocationStats(node.size(), forecastedWriteLoad, forecastedDiskUsage);
```

Comment:
It would be very useful to expose the amount of shards on undesired locations here as well, however that is not straightforward since desired balance is only kept in memory of the elected master and this stats are computed on every node locally.

There are 2 ways it could be resolved:
* request desired balance or a number of undesired shards per node from here. This would require multiple requests to the elected master
* Additional logic in https://github.com/elastic/elasticsearch/blob/7a3d58e7df2c943f974581dab511457f69e4e4bd/server/src/main/java/org/elasticsearch/action/admin/cluster/node/stats/TransportNodesStatsAction.java#L66-L69 to compute shard allocation stats. Since it relies on the cluster state and the transient desired balance it could be computed for all nodes at once on the coordinating node.

Reply:
++ to doing it once, see e.g. `org.elasticsearch.action.admin.cluster.stats.TransportClusterStatsAction#newResponseAsync` for prior art in this area.